[ML] Compute SHAP values for supervised learning

valeriy42 opened the pull request:
This PR introduces the computation of SHAP (SHapley Additive exPlanation) values for feature importance. Refer [Consistent Individualized Feature Attribution for Tree Ensembles](https://arxiv.org/abs/1802.03888) by Lundberg et al. for details to the original algorithm.

tveasey commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1445,5 +1446,54 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
+        std::make_unique<maths::CTreeShapFeatureImportance>(m_BestForest);
+    TDoubleVecVec shapValues;
+    TDoubleVec shapTotal;
+    size_t numberInputFields = predictionColumn(frame.numberColumns()) - 1;
```
Should this not be `m_Encoder->numberEncodedColumns() - 1`, i.e. the number of regressors? Also for consistency:
```suggestion
    std::size_t numberInputFields = predictionColumn(frame.numberColumns()) - 1;
```

valeriy42 replied:
Fixed both :+1:

tveasey commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1445,5 +1446,54 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
+    }
+    bool successful;
+    auto treeFeatureImportance =
+        std::make_unique<maths::CTreeShapFeatureImportance>(m_BestForest);
```
```suggestion
        std::make_unique<CTreeShapFeatureImportance>(m_BestForest);
```

valeriy42 replied:
fixed

tveasey commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1445,5 +1446,54 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
+        treeFeatureImportance->shap(frame, *m_Encoder, numberInputFields);
+
+    // get indices of the top elements
+    std::vector<std::size_t> indices(shapTotal.size());
```
I think we have the typedef
```suggestion
    TSizeVec indices(shapTotal.size());
```

valeriy42 replied:
Fixed

tveasey commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1445,5 +1446,54 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
+
+    // write top columns to the result
+    std::size_t offset{frame.numberColumns()};
+    frame.resizeColumns(m_NumberThreads, frame.numberColumns() + topShapValues);
```
Although these aren't computed until the end, my inclination is to only resize the data frame once and that this should therefore be reflected in the `numberExtraColumnsForTrain`. The resize operation is expensive and doing it twice means we have to copy the data again.

valeriy42 replied:
As discussed, this change would have trade-offs which are not necessarily advantageous.

tveasey commented on lib/api/CDataFrameTrainBoostedTreeRunner.cc:
```diff
@@ -296,6 +305,12 @@ const std::string CDataFrameTrainBoostedTreeRunner::FEATURE_BAG_FRACTION{"featur
 const std::string CDataFrameTrainBoostedTreeRunner::NUMBER_ROUNDS_PER_HYPERPARAMETER{"number_rounds_per_hyperparameter"};
 const std::string CDataFrameTrainBoostedTreeRunner::BAYESIAN_OPTIMISATION_RESTARTS{"bayesian_optimisation_restarts"};
+const std::string CDataFrameTrainBoostedTreeRunner::TOP_SHAP_VALUES{"top_shap_values"};
+const std::string CDataFrameTrainBoostedTreeRunner::SHAP_PREFIX{"shap_"};
```
Maybe
```suggestion
const std::string CDataFrameTrainBoostedTreeRunner::SHAP_PREFIX{"shap."};
```
This way they translated into nested fields and this matches outlier detection feature influence.

valeriy42 replied:
Fixed, also I moved the constant into `CDataFrameRegressionModel`

tveasey commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1445,5 +1446,54 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
+    frame.resizeColumns(m_NumberThreads, frame.numberColumns() + topShapValues);
+    TStrVec columnNames(frame.columnNames());
+    for (int i = 0; i < topShapValues; ++i) {
+        columnNames[offset + i] = SHAP_PREFIX + frame.columnNames()[indices[i]];
```
I can't see where this is defined, but we'll have a circular library dependency problem if this uses the version in `api` library. I'd be inclined to pass this in.

valeriy42 replied:
That is why I moved the definition to `CDataFrameRegressionModel` which is a part of `maths`

tveasey commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1445,5 +1446,54 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
+    std::tie(std::ignore, successful) = frame.writeColumns(
+        m_NumberThreads, 0, frame.numberRows(), [&](TRowItr beginRows, TRowItr endRows) {
+            for (auto row = beginRows; row != endRows; ++row) {
+                for (int i = 0; i < topShapValues; ++i) {
```
nit: can we make loops over `std::size_t` in this function.

valeriy42 replied:
Following your suggestion, I removed the need for the loops here altogether.

tveasey commented on lib/maths/CDataFrameCategoryEncoder.cc:
```diff
@@ -231,8 +231,12 @@ std::size_t CEncodedDataFrameRowRef::numberColumns() const {
+    : CDataFrameCategoryEncoder(&builder) {
+}
+
+CDataFrameCategoryEncoder::CDataFrameCategoryEncoder(CMakeDataFrameCategoryEncoder* builder) {
```
This seems odd to me: if this is passed in by pointer then it could be null, but this can't support it being null. What is forcing us to create this new constructor?

valeriy42 replied:
Hm, I need this to override the logic of `CMakeDataFrameCategoryEncoder` for unit testing. To this end, inherit from `CMakeDataFrameCategoryEncoder`, but if I pass it by reference to the encoder, I loose polymorphic behavior. Do you have any better idea?

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,248 @@
+                    phiVec.emplace_back(std::move(phi));
+                }
+            },
+            std::make_pair(TDoubleVecVec(), TDoubleVec(numberFeatures, 0))));
```
We should `reserve` this vector. However, I'm a bit uncomfortable about writing all these values to a vector of vectors and then writing a subset of them back to the data frame. This particularly breaks matters if the frame is stored off heap for size constraints. I feel like a better strategy is to resize the data frame first and write all the values directly into the data frame storage. When we come to write results we can simply write out the top n.

valeriy42 replied:
I fixed it by writing directly to the data frame.

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,248 @@
+    auto result = frame.readRows(
+        1, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](std::vector<double>& state, TRowItr beginRows, TRowItr endRows) {
```
```suggestion
            [&](TDoubleVec& state, TRowItr beginRows, TRowItr endRows) {
```

valeriy42 replied:
Fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    auto result = frame.readRows(
+        1, 0, frame.numberRows(),
```
I think this should be thread-able no? Basically, you get a copy of state for each thread so as long as you can combine and this isn't mutating any other state you should be ok.

valeriy42 replied:
Fixed.

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,248 @@
+std::size_t CTreeShapFeatureImportance::updateNodeValues(TTree& tree,
+                                                         std::size_t nodeIndex,
+                                                         const TDoubleVec& samplesPerNode,
+                                                         std::size_t depth) {
```
const or better static?

valeriy42 replied:
fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,248 @@
+                                               std::size_t nodeIndex,
+                                               double parentFractionZero,
+                                               double parentFractionOne,
+                                               int parentFeatureIndex) {
```
const or better static?

valeriy42 replied:
fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,248 @@
+void CTreeShapFeatureImportance::extendPath(SPath& path,
+                                            double fractionZero,
+                                            double fractionOne,
+                                            int featureIndex) {
```
const or better static?

valeriy42 replied:
fixed

valeriy42 commented on lib/api/CDataFrameTrainBoostedTreeRegressionRunner.cc:
```diff
@@ -73,6 +75,18 @@ void CDataFrameTrainBoostedTreeRegressionRunner::writeOneRow(
     writer.Key(IS_TRAINING_FIELD_NAME);
     writer.Bool(maths::CDataFrameUtils::isMissing(row[columnHoldingDependentVariable]) == false);
+    if (this->topShapValues() > 0) {
+        for (int i = 0; i < row.numberColumns(); ++i) {
```
Fixed in 6f81bdad

valeriy42 commented on lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:
```diff
@@ -593,7 +593,7 @@ BOOST_AUTO_TEST_CASE(testRunBoostedTreeRegressionTrainingWithStateRecovery) {
                 params.s_Lambda, params.s_Gamma, params.s_SoftTreeDepthLimit,
                 params.s_SoftTreeDepthTolerance, params.s_Eta, params.s_MaximumNumberTrees,
-                params.s_FeatureBagFraction, &persisterSupplier);
+                params.s_FeatureBagFraction, 0, &persisterSupplier);
```
fixed

valeriy42 commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,248 @@
+    return totalSamplesPerNode;
+}
+
+ml::maths::CTreeShapFeatureImportance::CTreeShapFeatureImportance(TTreeVec trees)
```
fixed

valeriy42 commented on lib/maths/CBoostedTreeFactory.cc:
```diff
@@ -1017,5 +1017,10 @@ void CBoostedTreeFactory::noopRecordProgress(double) {
 }
+
+CBoostedTreeFactory& CBoostedTreeFactory::shapValues(bool shapValues) {
+    m_TreeImpl->m_ShapValues = shapValues;
```
Fixed in 006698f5

valeriy42 commented on lib/api/CDataFrameTrainBoostedTreeRunner.cc:
```diff
@@ -157,6 +159,9 @@ CDataFrameTrainBoostedTreeRunner::CDataFrameTrainBoostedTreeRunner(
         m_BoostedTreeFactory->bayesianOptimisationRestarts(bayesianOptimisationRestarts);
     }
+    if (topShapValues > 0) {
+        m_TopShapValues = topShapValues;
```
Fixed in 87dfec17

valeriy42 commented:
Thank you @tveasey for your helpful reviews. I adjusted the code as you suggested. Now I am writing the SHAP values directly into the data frame instead of an extra array. I changed the memory estimation accordingly in the factory class, please let me know if this suffice our I need to adjust the impl class as well.

tveasey commented on lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:
```diff
@@ -817,4 +817,128 @@ BOOST_AUTO_TEST_CASE(testCategoricalFieldsEmptyAsMissing) {
+        rapidjson::ParseResult ok(results.Parse(output.str()));
+        BOOST_TEST_REQUIRE(static_cast<bool>(ok) == true);
+
+        double c1_sum, c2_sum, c3_sum, c4_sum;
```
nit: can we stick to coding standards, i.e. c1Sum, etc.

valeriy42 replied:
fixed

tveasey commented on lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:
```diff
@@ -817,4 +817,128 @@ BOOST_AUTO_TEST_CASE(testCategoricalFieldsEmptyAsMissing) {
+                c4_sum += std::fabs(
+                    result["row_results"]["results"]["ml"]["shap.c4"].GetDouble());
+            }
+        }
```
I think it might be useful to state the linear function we're fitting here to explain these inequalities, i.e. `50 c_1 + 150 c_2 + 50 c_3 - 50 c_4`. This leads to a question: why do we expect `c1_sum` > `c3_sum`, since the feature values come from the same distribution? Isn't this just an artefact of sampling, in which case could we maybe make w_3 < w_1 < w_2.

tveasey replied:
Ok, reading the code above more carefully, I think this comes from the fact that mean(|c1 feature values|) > mean(|c3 feature values|), etc. So I think the test is good as is, but this needs more explanation.

valeriy42 replied:
Doc adjusted.

tveasey commented on lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:
```diff
@@ -817,4 +817,128 @@ BOOST_AUTO_TEST_CASE(testCategoricalFieldsEmptyAsMissing) {
+        TDoubleVec values;
+        rng.generateUniformSamples(-10.0, 10.0, weights.size() * rows, values);
+
+        // make last column categorical
```
Isn't this making the first column categorical?

valeriy42 replied:
Good catch. The comment was outdated.

tveasey commented on lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:
```diff
@@ -817,4 +817,128 @@ BOOST_AUTO_TEST_CASE(testCategoricalFieldsEmptyAsMissing) {
+        rapidjson::ParseResult ok(results.Parse(output.str()));
+        BOOST_TEST_REQUIRE(static_cast<bool>(ok) == true);
+
+        double c1_sum, c2_sum, c3_sum, c4_sum;
```
Unused

valeriy42 replied:
fixed

tveasey commented on lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:
```diff
@@ -817,4 +817,128 @@ BOOST_AUTO_TEST_CASE(testCategoricalFieldsEmptyAsMissing) {
+    }
+
+    // test with 0 shap values
+    {
```
Could we maybe factor out boilerplate code to set up the input data as a lambda?

valeriy42 replied:
I created a fixture to deal with the boiler plate code

tveasey commented on lib/maths/unittest/CTreeShapFeatureImportanceTest.cc:
```diff
@@ -0,0 +1,189 @@
+            }
+        }
+    });
+}
```
I think a nice test would be to test invariant "sum of shap values" = "prediction - mean" for some randomised input.

valeriy42 replied:
Done.

tveasey commented on lib/maths/CBoostedTreeImpl.cc:
```diff
@@ -1445,5 +1451,50 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
+                         });
+        m_TopShapIndices = TSizeVec(topShapValues);
+        std::copy(indices.begin(), indices.begin() + topShapValues,
+                  m_TopShapIndices.get().begin());
```
I'm reading this as we output top overall shap values for each feature for each example. It seems to me like it would be more useful to return separate result: feature shap values, and then top (local) shap values for each example. If we return indices for all shap values then the writer can simply iterate over them and select the top n. You'd then introduce fields called `shap.<feature foo>`, so different results would have different fields set. WDYT?

valeriy42 replied:
I adjusted the code to output top shap values per row.

tveasey commented on include/maths/CBoostedTree.h:
```diff
@@ -391,6 +399,8 @@ class MATHS_EXPORT CBoostedTree final : public CDataFrameRegressionModel {
     //! Get the column containing the model's prediction for the dependent variable.
     std::size_t columnHoldingPrediction(std::size_t numberColumns) const override;
 
+    const TOptionalSizeVec& columnsHoldingShapValues() const override;
```
How about just returning an empty vector if there aren't anything rather than making this optional? Seems cleaner.

valeriy42 replied:
Indeed. However, this now becomes irrelevant as I let the result writer to pick top SHAP values for every row.

valeriy42 commented on include/maths/CDataFrameRegressionModel.h:
```diff
@@ -53,6 +62,14 @@ class MATHS_EXPORT CDataFrameRegressionModel {
 
+    //! Get the range of column indices with SHAP values
+    //!
+    //! \note The boost::irange can be sustituted by ranges::views::iota once this becomes available in C++20
```
Fixed

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+            : s_FractionOnes(length), s_FractionZeros(length),
+              s_FeatureIndex(length, -1), s_Scale(length), s_NextIndex(0),
+              s_MaxLength(length) {}
+        ~SPath() = default;
```
Good catch. I included it while hunting a memory leak, but it has become obsolete since.

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+    using TDoubleVecVec = std::vector<TDoubleVec>;
+
+    struct SPath {
+        SPath(std::size_t length)
```
Fixed

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+    //!\param[in] numberFeatures number of features. If set to -1, it's assument that the number of feature is equal to the
+    //! number of columns in the \p frame.
+    //!\return pair shap values for all points, sum of magnitues of shap values for every feature.
+    CTreeShapFeatureImportance::TDoubleVec shap(core::CDataFrame& frame,
```
Fixed

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+                           double fractionZero,
+                           double fractionOne,
+                           int featureIndex);
+    static double sumUnwoundPath(const CTreeShapFeatureImportance::SPath& path, int pathIndex);
```
fixed

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+                           double fractionOne,
+                           int featureIndex);
+    static double sumUnwoundPath(const CTreeShapFeatureImportance::SPath& path, int pathIndex);
+    static void unwindPath(CTreeShapFeatureImportance::SPath& path, int pathIndex);
```
fixed

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+
+    //! Compute SHAP values for the data in \p frame using the specified \p encoder.
+    //!
+    //!\param[in] numberFeatures number of features. If set to -1, it's assument that the number of feature is equal to the
```
fixed

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+    //!
+    //!\param[in] numberFeatures number of features. If set to -1, it's assument that the number of feature is equal to the
+    //! number of columns in the \p frame.
+    //!\return pair shap values for all points, sum of magnitues of shap values for every feature.
```
fixed

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+                   std::size_t numThreads);
+
+    //! Recursively computes inner node values as weighted average of the children (leaf) values
+    //! Returns maximum depth the the tree
```
Fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+                                 std::size_t numberFeatures,
+                                 std::size_t offset) {
+    numberFeatures = (numberFeatures != -1) ? numberFeatures : frame.numberColumns();
+    std::vector<std::size_t> maxDepthVec;
```
```suggestion
    TSizeVec maxDepthVec;
```

valeriy42 replied:
done

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+                                           const CDataFrameCategoryEncoder& encoder,
+                                           std::size_t numThreads) {
+    auto result = frame.readRows(
+        numThreads, 0, frame.numberRows(),
```
Also, defaults to range to `[0, frame.numberRows())` if not explicitly provided
```suggestion
       m_NumberThreads,
```

valeriy42 replied:
Fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+    maxDepthVec.reserve(m_Trees.size());
+    for (auto& tree : m_Trees) {
+        auto samplesPerNode =
+            CTreeShapFeatureImportance::samplesPerNode(tree, frame, encoder, 0);
```
This is passing in 0 for number of threads! But I think the method can just use the member variable in any case and we can drop this parameter.

valeriy42 replied:
Good catch!

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto node{tree[0]};
+                    state[0] += 1;
```
nit
```suggestion
                    state[0] += 1.0;
```

valeriy42 replied:
done

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+            [&](TDoubleVec& state, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto node{tree[0]};
```
I think we should make this type `const CBoostedTreeNode*` otherwise you are copying the entire node as you traverse the tree for every single row (and knock on changes below).
```suggestion
                    const CBoostedTreeNode* node{&tree[0]};
```

valeriy42 replied:
fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+    auto result = frame.readRows(
+        numThreads, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](TDoubleVec& state, TRowItr beginRows, TRowItr endRows) {
```
How about a more descriptive name?
```suggestion
            [&](TDoubleVec& samplesPerNode, TRowItr beginRows, TRowItr endRows) {
```

valeriy42 replied:
Fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+    double averageValue = (leftWeight * tree[node.leftChildIndex()].value() +
+                           rightWeight * tree[node.rightChildIndex()].value()) /
+                          (leftWeight + rightWeight);
+    node.value(averageValue);
```
👍 I like the idea of repurposing the internal node values for this! It is useful information to surface in its own right.

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+        return 0;
+    }
+
+    std::size_t depthLeft = CTreeShapFeatureImportance::updateNodeValues(
```
nit: can we stick to braces initialisation.

valeriy42 replied:
changed here and below.

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    auto result = frame.writeColumns(
+        m_NumberThreads, 0, frame.numberRows(),
```
Not needed since this is default behaviour.
```suggestion
        m_NumberThreads,
```

valeriy42 replied:
done

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+    auto result = frame.writeColumns(
+        m_NumberThreads, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](auto& state, TRowItr beginRows, TRowItr endRows) {
```
```suggestion
            [&](TDoubleVec& phiSum, TRowItr beginRows, TRowItr endRows) {
```

valeriy42 replied:
Fixed

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+            [&](auto& state, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto& phiSum = state;
```
We don't really need reference to argument
```suggestion
```

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+                        CTreeShapFeatureImportance::shapRecursive(
+                            m_Trees[i], m_SamplesPerNode[i], encoder,
+                            encodedRow, path, 0, 1.0, 1.0, -1, offset, row);
+                        i += 0;
```
```suggestion
```

valeriy42 replied:
whoopsy

tveasey commented on lib/maths/CTreeShapFeatureImportance.cc:
```diff
@@ -0,0 +1,258 @@
+                    }
+                    for (int j = 0; j < numberFeatures; ++j) {
+                        row->writeColumn(offset + j,
+                                         (*row)[offset + j] / m_Trees.size());
```
Can we pass in the normalisation factor and write once inside `shapRecursive`? I found this line confusing until I saw you were first writing them inside `shapRecursive`.

valeriy42 replied:
Done

valeriy42 commented on include/maths/CTreeShapFeatureImportance.h:
```diff
@@ -0,0 +1,138 @@
+                                                std::size_t numberFeatures,
+                                                std::size_t offset);
+
+    static CTreeShapFeatureImportance::TDoubleVec
```
fixed

valeriy42 commented on lib/api/CDataFrameTrainBoostedTreeRegressionRunner.cc:
```diff
@@ -18,6 +18,8 @@
 #include <api/CDataFrameAnalysisSpecification.h>
 #include <api/ElasticsearchStateIndex.h>
 
+#include <numeric>
```
done

tveasey commented on lib/api/CDataFrameTrainBoostedTreeClassifierRunner.cc:
```diff
@@ -153,6 +153,20 @@ void CDataFrameTrainBoostedTreeClassifierRunner::writeOneRow(
     }
+
+    if (this->topShapValues() > 0) {
+        auto shapColumnsRange{this->boostedTree().columnsHoldingShapValues()};
```
At least in theory `nth_element` needs to guard against the case `topShapValues() > shapColumnsRange.size()` unless that happens elsewhere? (Also, we probably don't need to actually copy the size vector; we could have a begin and end or hold the vector and return by reference, at which point it is more efficient to have a heap of the top values.) We have a class which makes all this easy in the `maths` library; I think using this works out as clean in code. You'd need a factory method to take care of the type of the comparison operator, but something like should work:
```
template<typename LESS>
auto makeLargestShapAccumulator(std::size_t n, LESS less) {
    return maths::CBasicStatistics::COrderStatisticsHeap<std::size_t, LESS>{n, std::size_t{}, less);
}
...
    auto largestShapValues = makeLargestShapAccumulator(this->topShapValues(), [&row](std::size_t lhs, std::size_t rhs) {
                             return std::fabs(row[lhs]) > std::fabs(row[rhs]);
                         });
    largestShapValues.add(this->boostedTree().columnsHoldingShapValues());
    largestShapValues.sort();
    for (auto i : largestShapValues) {
       ...
    }
```

valeriy42 replied:
Done. Great idea!

tveasey commented on lib/maths/CBoostedTreeFactory.cc:
```diff
@@ -1017,5 +1022,11 @@ void CBoostedTreeFactory::noopRecordProgress(double) {
 }
+
+CBoostedTreeFactory& CBoostedTreeFactory::topShapValues(std::size_t topShapValues) {
+    m_TopShapValues = topShapValues;
```
Do we need this variable on the factory as well: seems redundant?

valeriy42 replied:
I need it to estimate memory usage with resized data frame.

tveasey commented on lib/maths/CDataFrameCategoryEncoder.cc:
```diff
@@ -231,8 +231,12 @@ std::size_t CEncodedDataFrameRowRef::numberColumns() const {
 
-CDataFrameCategoryEncoder::CDataFrameCategoryEncoder(CMakeDataFrameCategoryEncoder builder) {
-    m_Encodings = builder.makeEncodings();
+CDataFrameCategoryEncoder::CDataFrameCategoryEncoder(CMakeDataFrameCategoryEncoder& parameters) {
```
nit: for consistency can we call this `builder`.

valeriy42 replied:
done

valeriy42 commented:
This PR introduces the computation of SHAP (SHapley Additive exPlanation) values for feature importance. Refer "Consistent Individualized Feature Attribution for Tree Ensembles" by Lundberg et al. for details to the original algorithm.
